### PR TITLE
copy the Address.toKey method off into its own class, add a deseriali…

### PR DESCRIPTION
**Pull request by AWinterman**
…zation method for the address.

### CODE REVIEW

File: src/main/java/com/urbanairship/datacube/OffsetBasedAddressFormat.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+    @Override
+    public Optional<Address> fromKey(byte[] bytes) throws IOException, InterruptedException {
+        int remainingBytes = bytes.length;
+
+        ByteBuffer bb = ByteBuffer.wrap(bytes);
+
+        if (useAddressPrefixByteHash) {
+            // then read the first int and throw it out.
+            bb.getInt();
```
Comment (AWinterman) [inline src/main/java/com/urbanairship/datacube/OffsetBasedAddressFormat.java]: todo: might be cool to a) include the partition on the built address, b) compute the hashcode again on read to make sure we haven't got data corruption.

File: CONTRIBUTORS [hunk 2]
```diff
@@ -3,3 +3,6 @@ Nate Putnam <nate@urbanairship.com>
 Gavin McQuillan <gavin@urbanairship.com>
 Tim Robertson <timrobertson100@gmail.com>
 Max Jakob <max.jakob@neofonie.de>
+Andrew Winterman <andrew.winterman@urbanairship.com>
+Cory Kolbeck <cbeck@urbanairship.com>
+Jeff Meyers <jeff.meyers@urbanairship.com>
```
Comment (todd534) [inline CONTRIBUTORS]: I guess I'm fucken chopped liver? 

(jkjkjk)

Comment (AWinterman) [inline CONTRIBUTORS]: I FORGOT TODD

File: src/main/java/com/urbanairship/datacube/OffsetBasedAddressFormat.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+                Optional<byte[]> maybeBucket = idService.getValueForId(dimensions.indexOf(dimension), bucket);
+                if (!maybeBucket.isPresent()) {
+                    return Optional.empty();
+                }
+                bucket = maybeBucket.get();
+            }
+
+            BucketType bucketType = null;
+            for (BucketType candidate : dimension.getBucketer().getBucketTypes()) {
+                if (Arrays.equals(candidate.getUniqueId(), bucketPrefix)) {
```
Comment (todd534) [inline src/main/java/com/urbanairship/datacube/OffsetBasedAddressFormat.java]: this strikes me as a wild ride, but I can't actually come up with a problem with it.

File: CONTRIBUTORS (hunk 2 above)
Comment (todd534) [reply CONTRIBUTORS]: you saw me commenting on this didn't u

File: src/main/java/com/urbanairship/datacube/Op.java [hunk 4]
```diff
@@ -6,16 +6,18 @@
...
 
+import java.util.function.Supplier;
+
 
 /**
  * A cell mutation or cell bucket. For example, a cube storing counters would contain Ops that
  * are numbers (see e.g. {@link LongOp}).
  */
-public interface Op extends CSerializable {
+public interface Op<T> extends CSerializable, Supplier<T> {
```
Comment (AWinterman) [inline src/main/java/com/urbanairship/datacube/Op.java]: this seems like an easy win to me-- you gain guarantees that you are combining compatible ops.

File: CONTRIBUTORS (hunk 2 above)
Comment (AWinterman) [reply CONTRIBUTORS]: haha

File: src/main/java/com/urbanairship/datacube/OffsetBasedAddressFormat.java (hunk 3 above)
Comment (AWinterman) [reply src/main/java/com/urbanairship/datacube/OffsetBasedAddressFormat.java]: what's wild about it? how could i make it less wild?

Comment (todd534) [reply src/main/java/com/urbanairship/datacube/OffsetBasedAddressFormat.java]: Can't think of a great way. I guess maybe move it into Bucketer would be my only idea? Anyway it's fine, I'm just being a jerk.

**Comment by AWinterman** [comment]
@todd534 👍 , although i do think that the lack of this functionality constrains what we can do and makes datacube seem perhaps more mystical than it should.

**Comment by AWinterman** [comment]
This is currently incorrect-- it needs to prepend the unique cube name when it attempts to deserialize (and the api should change so that the format prepends the cube name as well, and then we should stop doing so in the hbase db harness)

**Comment by AWinterman** [comment]
closing this thing cause lol it's been years

**Pull request closed without merging**